suite/tests/linux/rseq.c

```diff
@@ -812,6 +812,126 @@ test_rseq_native_abort(void)
 #endif /* DEBUG */
 }
 
+/* Tests that DR handles an asynch signal in the native code in the final rseq
+ * fragment.  We use a system call, which is officially disallowed.  We have special
+ * exceptions in the code which look for the test name "linux.rseq" and are limited to
+ * DEBUG.
+ */
+static void
+test_rseq_asynch_signal(void)
+{
+    volatile struct rseq *reg_rseq = get_my_rseq();
+#ifdef DEBUG /* See above: special code in core/ is DEBUG-only> */
+    int restarts = 0;
+#    ifdef X86
+    __asm__ __volatile__(
+        /* clang-format off */ /* (avoid indenting next few lines) */
+        RSEQ_ADD_TABLE_ENTRY(asynch_signal, 2f, 3f, 4f)
+        /* clang-format on */
+
+        "6:\n\t"
+        /* Store the entry into the ptr. */
+        "leaq rseq_cs_abort(%%rip), %%rax\n\t"
+        "movq %%rax, %[rseq_cs]\n\t"
+        "pxor %%xmm0, %%xmm0\n\t"
+        "mov $1,%%rcx\n\t"
+        "movq %%rcx, %%xmm1\n\t"
+
+        /* Restartable sequence. */
+        "2:\n\t"
+        /* Increase xmm0 every time.  DR (currently) won't restore xmm inputs
+         * to rseq sequences, nor does it detect that it needs to.
+         */
+        "paddq %%xmm1,%%xmm0\n\t"
+        "movq %%xmm0, %%rax\n\t"
+        /* We raise the signal on the instrumented run and never come back. */
```

Comment:
Add "in the instrumented run" to the test call-site in `main` and `instru` to the test function name.

Edit: oh so this test actually does the same thing for the native and instrumented run.

What about a case where there's a signal in the instrumented run but the native run completes properly? (I assume we indeed do a native run even if the instrumented one is interrupted by a signal or a side exit, right?)

Reply:
I'm trying to understand whether this can be a source of transparency violations: do rseq aborts from signals become significantly more likely during instrumented runs? Is it likely to have a degenerate case where the rseq region is never able to complete instrumented execution, because of repeated rseq aborts caused by signals during instrumented run?